Use Spring's ObjectMapper

dacrome opened the pull request.

dacrome committed:
Use Spring's ObjectMapper

tkrille commented:
Commit message: "Use Spring's ObjectMapper"

tkrille commented:
ITs are failing. Please revise your PR.

dacrome commented:
Updated commit message. On my machine all ITs are green. Which tests fail?

tkrille commented:
I'll re-test.

tkrille commented:
I've attached the test reports:

[org.osiam.client.AdminScopeIT.txt](https://github.com/osiam/osiam/files/24966/org.osiam.client.AdminScopeIT.txt)
[org.osiam.client.EditGroupServiceIT.txt](https://github.com/osiam/osiam/files/24968/org.osiam.client.EditGroupServiceIT.txt)
[org.osiam.client.SearchGroupServiceIT.txt](https://github.com/osiam/osiam/files/24967/org.osiam.client.SearchGroupServiceIT.txt)
[org.osiam.test.integration.SearchGroupsWithSortByIT.txt](https://github.com/osiam/osiam/files/24970/org.osiam.test.integration.SearchGroupsWithSortByIT.txt)

Seems that somethings wrong with groups.

tkrille commented:
ITs still failing :(

tkrille commented:
Can someone else please run the ITs? @wallner @sputnik27 @woglinde?

tkrille commented:
Tests are failing on the command line but are green in IntelliJ.

dacrome commented:
Strange. Any exceptions in the server log?

tkrille commented:
A lot, but it's hard to correlate. I'll run just the failed tests to see more. On what machines did you test? Vagrant VM?

dacrome commented:
How about just run a specific test and then check the logs?
`mvn clean verify -Dit.test=AdminScopeIT#can_search_for_groups`

> On what machines did you test? Vagrant VM?

With docker-machine/boot2docker 1.8.2 and in the vagrant vm with docker 1.7.2

tkrille commented:
> How about just run a specific test and then check the logs?

That's exactly what I was going to do.

tkrille commented on src/main/java/org/osiam/metrics/controller/MetricsController.java:
```diff
@@ -64,7 +67,7 @@
         final TimeUnit rateUnit = parseTimeUnit(RATE_UNIT, TimeUnit.SECONDS);
         final TimeUnit durationUnit = parseTimeUnit(DURATION_UNIT, TimeUnit.SECONDS);
-        return new ObjectMapper().registerModule(new MetricsModule(rateUnit, durationUnit, false));
+        return mapper.registerModule(new MetricsModule(rateUnit, durationUnit, false));
```
This will be called every time someone requests the metrics and will fail if it's called the second time. You cannot add the same module twice.

dacrome replied:
Ups. Yes, the module should only be registered once.

tkrille commented:
Unfortunately, running a single IT on the command line yields a green build. But I've found this:

```
2015-11-04 21:55:32.926  WARN 1 --- [io-8180-exec-16] o.o.r.helper.AttributesRemovalHelper     : Unable to serialize search result

com.fasterxml.jackson.databind.JsonMappingException: Can not resolve PropertyFilter with id 'filter properties by name'; no FilterProvider configured (through reference chain: java.util.ArrayList[0])
    at com.fasterxml.jackson.databind.ser.std.StdSerializer.findPropertyFilter(StdSerializer.java:274)
    at com.fasterxml.jackson.databind.ser.std.BeanSerializerBase.serializeFieldsFiltered(BeanSerializerBase.java:678)
    at com.fasterxml.jackson.databind.ser.BeanSerializer.serialize(BeanSerializer.java:150)
    at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:100)
    at com.fasterxml.jackson.databind.ser.impl.IndexedListSerializer.serializeContents(IndexedListSerializer.java:21)
    at com.fasterxml.jackson.databind.ser.std.AsArraySerializerBase.serialize(AsArraySerializerBase.java:183)
    at com.fasterxml.jackson.databind.ser.DefaultSerializerProvider.serializeValue(DefaultSerializerProvider.java:128)
    at com.fasterxml.jackson.databind.ObjectWriter._configAndWriteValue(ObjectWriter.java:800)
    at com.fasterxml.jackson.databind.ObjectWriter.writeValueAsString(ObjectWriter.java:676)
    at org.osiam.resources.helper.AttributesRemovalHelper.getJsonResponseWithAdditionalFields(AttributesRemovalHelper.java:117)
    at org.osiam.resources.helper.AttributesRemovalHelper.removeSpecifiedAttributes(AttributesRemovalHelper.java:65)
    at org.osiam.resources.controller.GroupController.searchWithGet(GroupController.java:128)
        ...
```

tkrille commented:
It's definitively the aforementioned exception with the same error message, that causes the trouble.

dacrome commented:
Blocked by #134

dacrome commented:
Rebased and updated. But some ITs are failing:

`Tests run: 467, Failures: 5, Errors: 0, Skipped: 8`

```
Failed tests: 
  SearchUserServiceIT.search_for_user_with_just_some_fields:326 
Expected: null
     but: was <Name [formatted=Mr. Gregg Barker, familyName=Barker, givenName=Gregg, middleName=null, honorificPrefix=null, honorificSuffix=null]>
  SearchUserServiceIT.search_for_user_with_some_attributes:338 
Expected: is <0>
     but: was <1>
  ControllerIT.REGT-001-#testCase: An API request missing an accept header with scope #scope and content type #contentType on path #requestPath should return HTTP status code #expectedResponseCode and content type #expectedResponseType.:76 Condition not satisfied:

responseStatusCode == expectedResponseCode
|                  |  |
409                |  200
                   false

  ControllerIT.REGT-001-#testCase: An API request missing an accept header with scope #scope and content type #contentType on path #requestPath should return HTTP status code #expectedResponseCode and content type #expectedResponseType.:76 Condition not satisfied:

responseStatusCode == expectedResponseCode
|                  |  |
409                |  200
                   false

  ScimExtensionTypesIT.URI: /osiam/extension-definition will return all persisted extension definitions:112 Condition not satisfied:

statusCode == 200
|          |
409        false
```

Can someone run the ITs and confirm this?

dacrome commented:
I'm not happy with the [mixin](https://github.com/osiam/osiam/pull/54/files#diff-7034ae061f8b76ce47e9e0b169b80651R161). I think about using the [`@JsonFilter`](http://wiki.fasterxml.com/JacksonFeatureJsonFilter), but then I need to add all `ObjectMapper`s in the unit tests a filter named `filter properties by name`.

dacrome commented:
Updated. But I can still not [inject the object mapper](https://github.com/osiam/osiam/pull/54/files#diff-7034ae061f8b76ce47e9e0b169b80651R55) without failing ITs.

timopick commented:
@dacrome Jackson's ObjectMapper is no Spring bean, so it can't, right?

timopick commented:
I've found [this answer](http://stackoverflow.com/a/30259658) while googling. They built a factory around it.

dacrome commented:
It is [created by Spring Boot](http://docs.spring.io/spring-boot/docs/current-SNAPSHOT/reference/htmlsingle/#howto-customize-the-jackson-objectmapper) by default as bean.

dacrome commented:
Actually it's not a problem with the Spring managed object mapper, but to add a mixin and filter to a application wide used object mapper because of side effects.

We want to get ride off the [AttributesRemovalHelper](https://github.com/dacrome/osiam/blob/using-springs-object-mapper/src/main/java/org/osiam/resources/helper/AttributesRemovalHelper.java) because we should do this already when fetching the data from database, so I think my current solution is okay: The [AttributesRemovalHelper](https://github.com/osiam/osiam/pull/54/files#diff-7034ae061f8b76ce47e9e0b169b80651R39) is now a Bean and only creates the object mapper once, so it's better like before to create an object mapper per request.

wallner commented on src/main/java/org/osiam/resources/controller/GroupController.java:
```diff
@@ -51,7 +51,8 @@
 
-    private AttributesRemovalHelper attributesRemovalHelper = new AttributesRemovalHelper();
+    @Autowired
+    private AttributesRemovalHelper attributesRemovalHelper;
```
Can we switch to constructor injection here too, please?

wallner commented on src/main/java/org/osiam/resources/controller/UserController.java:
```diff
@@ -61,7 +67,8 @@
 
-    private AttributesRemovalHelper attributesRemovalHelper = new AttributesRemovalHelper();
+    @Autowired
+    private AttributesRemovalHelper attributesRemovalHelper;
```
Please use constructor injection.

dacrome replied:
Not with this PR. I will create an own PR for complete switch to constructor injection.

wallner replied:
Let's do it with this one and be done with it.

wallner replied:
Ok, if you don't forget it.

dacrome replied:
Nope: #144

wallner commented:
Thank you.
